### PR TITLE
Prefer using PKG-INFO from .egg-info in assemble #3083

### PR DESCRIPTION
**Pull request by JonoYang**
This PR addresses #3083, where we have no dependencies returned when we scanned an installable Python codebase. This issue is cause by `BaseExtractedPythonLayout.assemble()`, where it would use the package data from the PKG-INFO file from the root of a Python codebase rather than from the PKG-INFO file from the .egg-info directory (also located in the root of a Python codebase). 

We want to use the PKG-INFO file from the .egg-info directory because the package data from that contains the dependency information.

### CODE REVIEW

File: tests/packagedcode/data/pypi/unpacked_sdist/prefer-egg-info-pkg-info/celery-expected.json [hunk 1]
```diff
@@ -0,0 +1,8105 @@
...
+              "is_runtime": true,
+              "is_optional": false,
+              "is_resolved": false,
+              "resolved_package": {},
+              "extra_data": {}
+            },
+            {
+              "purl": "pkg:pypi/azure-storage-blob@12.9.0",
+              "extracted_requirement": "azure-storage-blob==12.9.0; extra == \"azureblockblob\"",
+              "scope": "azurebl
```
Comment (pombredanne) [inline tests/packagedcode/data/pypi/unpacked_sdist/prefer-egg-info-pkg-info/celery-expected.json]: We should have an extracted requirement of "msgpack"

Comment (pombredanne) [inline tests/packagedcode/data/pypi/unpacked_sdist/prefer-egg-info-pkg-info/celery-expected.json]: We should have instead:
```suggestion
              "extracted_requirement": "couchbase>=3.0.0",
```

File: tests/packagedcode/data/pypi/unpacked_sdist/prefer-egg-info-pkg-info/celery-expected.json [hunk 2]
```diff
@@ -0,0 +1,8105 @@
...
+      "for_package_uid": "pkg:pypi/celery@5.2.7?uuid=fixed-uid-done-for-testing-5642512d1758",
+      "datafile_path": "celery/celery.egg-info/PKG-INFO",
+      "datasource_id": "pypi_editable_egg_pkginfo"
+    },
+    {
+      "purl": "pkg:pypi/pylibmc",
+      "extracted_requirement": "pylibmc; platform_system != \"Windows\" and extra == \"memcache\"",
+      "scope": "memcache",
+      "is_runtime": true,
+      "is_optional": false,
```
Comment (pombredanne) [inline tests/packagedcode/data/pypi/unpacked_sdist/prefer-egg-info-pkg-info/celery-expected.json]: IMHO we should be true for "extras" e.g. not an "install_requires" and therefore not in "install" scope:
```suggestion
      "is_optional": true,
```

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks!
There are a few nit that likely are leftover from before... we should better track the extracted requirements and treat PyPI extras as optional

**Comment by JonoYang** [comment]
I've updated the code to return the dependency name in the extracted_requirements and to properly set the is_optional field on the dependencies listed in extra_requires.

Tests are passing, so I will merge this.

**Pull request merged**